Release/src/http/client/winhttppal/winhttppal.cpp

```diff
@@ -0,0 +1,3804 @@
+    }
+
+    void ResumeTransfer(CURL *handle, int bitmask)
+    {
+        int still_running;
+
+        m_MultiMutex.lock();
+        curl_easy_pause(handle, bitmask);
+        curl_multi_perform(m_curlm, &still_running);
+        m_MultiMutex.unlock();
+    }
+
+    BOOL AddHandle(std::shared_ptr<WinHttpRequestImp> srequest, CURL *handle)
+    {
+        CURLMcode mres = CURLM_OK;
+
+        m_MultiMutex.lock();
+        if (std::find(m_ActiveCurl.begin(), m_ActiveCurl.end(), handle) != m_ActiveCurl.end()) {
+            mres = curl_multi_remove_handle(m_curlm, handle);
+            m_ActiveCurl.erase(std::remove(m_ActiveCurl.begin(), m_ActiveCurl.end(), handle), m_ActiveCurl.end());
+        }
+        if (std::find(m_ActiveRequests.begin(), m_ActiveRequests.end(), srequest) != m_ActiveRequests.end()) {
+            m_ActiveRequests.erase(std::remove(m_ActiveRequests.begin(), m_ActiveRequests.end(), srequest), m_ActiveRequests.end());
+        }
+        m_MultiMutex.unlock();
+        if (mres != CURLM_OK)
+        {
+            TRACE("curl_multi_remove_handle() failed: %s\n", curl_multi_strerror(mres));
+            return FALSE;
+        }
+
+        m_MultiMutex.lock();
+        mres = curl_multi_add_handle(m_curlm, handle);
+        m_ActiveCurl.push_back(handle);
+        m_ActiveRequests.push_back(srequest);
+        m_MultiMutex.unlock();
+        if (mres != CURLM_OK)
+        {
+            TRACE("curl_multi_add_handle() failed: %s\n", curl_multi_strerror(mres));
+            return FALSE;
+        }
+
+        return TRUE;
+    }
+    BOOL RemoveHandle(std::shared_ptr<WinHttpRequestImp> srequest, CURL *handle, bool clearPrivate)
+    {
+        CURLMcode mres;
+
+        m_MultiMutex.lock();
+        mres = curl_multi_remove_handle(m_curlm, handle);
+
+        if (clearPrivate)
+            curl_easy_getinfo(handle, CURLINFO_PRIVATE, NULL);
+
+        m_ActiveCurl.erase(std::remove(m_ActiveCurl.begin(), m_ActiveCurl.end(), handle), m_ActiveCurl.end());
+        m_ActiveRequests.erase(std::remove(m_ActiveRequests.begin(), m_ActiveRequests.end(), srequest), m_ActiveRequests.end());
+        m_MultiMutex.unlock();
+        if (mres != CURLM_OK)
+        {
+            TRACE("curl_multi_add_handle() failed: %s\n", curl_multi_strerror(mres));
+            return FALSE;
+        }
+
+        return TRUE;
+    }
+
+    long GetTimeout()
+    {
+        long curl_timeo;
+
+        m_MultiMutex.lock();
+        curl_multi_timeout(m_curlm, &curl_timeo);
+        m_MultiMutex.unlock();
+
+        return curl_timeo;
+    }
+
+    void KickStart()
+    {
+        std::lock_guard<std::mutex> lck(m_hAsyncEventMtx);
+        m_hAsyncEventCounter++;
+        m_hAsyncEvent.notify_all();
+    }
+
+    ComContainer(): m_hAsyncEventCounter(0)
+    {
+        curl_global_init(CURL_GLOBAL_ALL);
+        thread_setup();
+
+        m_curlm = curl_multi_init();
+        
+        DWORD thread_id;
+        m_hAsyncThread = CREATETHREAD(
```

Comment:
Hmmm I think this should respect the scheduler instead of spinning up its own thread?

Reply:
Also why not just use std::thread?

Reply:
This one is a long story but I'm happy to share. I was initially using std::thread. We found out that We are leaking openssl context for each thread used on openssl 1.1 because pthread_once() doesn't work for std::thread created threads.

Apparently, OpenSSL allocates data structures and stores them into the thread local storage (TLS). The clean up of this data structure and one time allocation only works if the thread was created by pthread or windows API.

This took me one week to figure out.